### DISCUSSION TITLE
Deserializing pointers in structures that form graphs

### DISCUSSION THREAD

**Discussion by vibraphone** [nlohmann/json#2916] [Q&A]
Is there any utility provided by nlohmann/json for [pointer swizzling](https://en.wikipedia.org/wiki/Pointer_swizzling)? The objects I'm serializing/deserializing have pointers to like instances that can form cycles, so there's not necessarily an ordering that guarantees pointers will be available as objects are deserialized.

The problem I've run into while implementing swizzling is that emplacing/moving objects means the swizzler cannot hold on to the address of a pointer that needs to be filled in when the referenced object is deserialized.

---

### Example

Consider structs of type `foo` with instances A and B. Each has a member pointer `bar` that references the other. Say A is deserialized first. Its pointer (`A.bar`) to B cannot be populated until B is deserialized. But if A is constructed on the heap and then moved/emplaced into a std container, then `&A.bar` in `from_json()` points to the heap, not the final destination of A. 

---

So, the strategy of holding references to the addresses member variables that need to be updated fails. Has anyone else found a solution to this issue? None of the persistent stores I'm familiar with (e.g. [Texas](https://link.springer.com/chapter/10.1007/978-1-4471-3209-7_2)) seem to use or accommodate modern C++ features.

The only solution I can think of is to use a special class (say `swizzle<T>`) that behaves like a `std::weak-pointer<T>` in reverse – it is null until its referenced object comes into existence. This would be pretty intrusive.

**Comment by nlohmann** [comment]
nlohmann/json has no built-in pointer-swizzling or reference-resolution mechanism; it is a DOM library with to_json/from_json hooks and has no concept of object identity across the tree. The standard workaround for cyclic/graph-shaped data is a two-pass, ID-based approach rather than address-based swizzling: (1) give each node a stable ID field in the JSON, (2) in a first pass deserialize all nodes into a container (e.g. std::map<ID, std::shared_ptr<Node>> or a std::deque so addresses stay stable) without resolving pointer members, storing the referenced IDs instead, and (3) in a second pass walk the structures and replace the ID placeholders with the actual shared_ptr/weak_ptr looked up in the registry. This avoids the address-after-move problem you ran into, since resolution happens only after every object has its final, stable storage location. There is no simpler answer within the library itself for this use case.

---
*This reply was drafted by [Claude Code](https://claude.com/claude-code) on behalf of @nlohmann, as part of a review of open discussions.*
